pkg/fuzzer/fuzzer.go

```diff
@@ -0,0 +1,307 @@
+}
+
+type Request struct {
+	Prog         *prog.Prog
+	NeedCover    bool
+	NeedRawCover bool
+	NeedSignal   bool
+	NeedHints    bool
+	// Fields that are only relevant within pkg/fuzzer.
+	flags   ProgTypes
+	stat    string
+	result  *Result
+	resultC chan *Result
+}
+
+type Result struct {
+	Info *ipc.ProgInfo
+	Stop bool
+}
+
+func (fuzzer *Fuzzer) Done(req *Request, res *Result) {
+	// Triage individual calls.
+	// We do it before unblocking the waiting threads because
+	// it may result it concurrent modification of req.Prog.
+	if req.NeedSignal && res.Info != nil {
+		for call, info := range res.Info.Calls {
+			fuzzer.triageProgCall(req.Prog, &info, call, req.flags)
+		}
+		fuzzer.triageProgCall(req.Prog, &res.Info.Extra, -1, req.flags)
+	}
+	// Unblock threads that wait for the result.
+	req.result = res
+	if req.resultC != nil {
+		req.resultC <- res
+	}
+	// Update stats.
+	fuzzer.mu.Lock()
+	fuzzer.stats[req.stat]++
+	delete(fuzzer.runningExecs, req)
+	fuzzer.mu.Unlock()
+}
+
+func (fuzzer *Fuzzer) triageProgCall(p *prog.Prog, info *ipc.CallInfo, call int,
+	flags ProgTypes) {
+	prio := signalPrio(p, info, call)
+	if !fuzzer.Corpus.AddRawMaxSignal(info.Signal, prio) {
+		return
+	}
+	fuzzer.Logf(2, "found new signal in call %d in %s\n",
+		call, string(p.Serialize()))
+	fuzzer.startJob(&triageJob{
+		p:     p.Clone(),
+		call:  call,
+		info:  *info,
+		flags: flags,
+	})
+}
+
+func signalPrio(p *prog.Prog, info *ipc.CallInfo, call int) (prio uint8) {
+	if call == -1 {
+		return 0
+	}
+	if info.Errno == 0 {
+		prio |= 1 << 1
+	}
+	if !p.Target.CallContainsAny(p.Calls[call]) {
+		prio |= 1 << 0
+	}
+	return
+}
+
+type Candidate struct {
+	Prog      *prog.Prog
+	Hash      hash.Sig
+	Smashed   bool
+	Minimized bool
+}
+
+func (fuzzer *Fuzzer) NextInput() *Request {
+	req := fuzzer.nextInput()
+	fuzzer.mu.Lock()
+	fuzzer.runningExecs[req] = time.Now()
+	fuzzer.mu.Unlock()
+	if req.stat == statCandidate {
+		if fuzzer.queuedCandidates.Load() <= 0 {
+			panic("queuedCandidates is out of sync")
+		}
+		fuzzer.queuedCandidates.Add(^uint64(0))
+	}
+	if fuzzer.Config.NeedCandidates != nil &&
+		fuzzer.NeedCandidates() &&
+		!fuzzer.candidatesRequested.CompareAndSwap(false, true) {
+		select {
+		case fuzzer.Config.NeedCandidates <- struct{}{}:
+		default:
+		}
+	}
+	return req
+}
+
+func (fuzzer *Fuzzer) nextInput() *Request {
+	nextExec := fuzzer.nextExec.tryPop()
+	if nextExec != nil {
+		return nextExec.value
+	}
+	// Either generate a new input or mutate an existing one.
+	mutateRate := 0.95
+	if !fuzzer.Config.Coverage {
+		// If we don't have real coverage signal, generate programs
+		// more frequently because fallback signal is weak.
+		mutateRate = 0.5
+	}
+	rnd := fuzzer.rand()
+	if rnd.Float64() < mutateRate {
+		req := mutateProgRequest(fuzzer, rnd)
+		if req != nil {
+			return req
+		}
+	}
+	return genProgRequest(fuzzer, rnd)
+}
+
+func (fuzzer *Fuzzer) startJob(newJob job) {
+	fuzzer.Logf(1, "started %T", newJob)
```

Comment:
Log level 1 will surface in syz-manager web log. Do we want every job there? Looks too fine-grained.

Reply:
If you mean syz-manager itself, then no, it won't be sent via RPC.

Reply:
I meant when we move pkg/fuzzer into syz-manager.

Reply:
Ah, I see. I've set it to 2.